File: backend/app/admin/service/internal/service/menu.go
package service

import (
	"context"

	"github.com/go-kratos/kratos/v2/log"
	"google.golang.org/protobuf/types/known/emptypb"

	"kratos-monolithic-demo/app/admin/service/internal/data"

	adminV1 "kratos-monolithic-demo/gen/api/go/admin/service/v1"
	"kratos-monolithic-demo/gen/api/go/common/pagination"
	systemV1 "kratos-monolithic-demo/gen/api/go/system/service/v1"
)

type MenuService struct {
	adminV1.MenuServiceHTTPServer

	log *log.Helper

	uc *data.MenuRepo
}

func NewMenuService(uc *data.MenuRepo, logger log.Logger) *MenuService {
	l := log.NewHelper(log.With(logger, "module", "menu/service/admin-service"))
	return &MenuService{
		log: l,
		uc:  uc,
	}
}

func (s *MenuService) ListMenu(ctx context.Context, req *pagination.PagingRequest) (*systemV1.ListMenuResponse, error) {
	ret, err := s.uc.List(ctx, req)
	if err != nil {

		return nil, err
	}

	return ret, nil
}

func (s *MenuService) GetMenu(ctx context.Context, req *systemV1.GetMenuRequest) (*systemV1.Menu, error) {
	ret, err := s.uc.Get(ctx, req)
	if err != nil {

		return nil, err
	}

	return ret, nil
}

func (s *MenuService) CreateMenu(ctx context.Context, req *systemV1.CreateMenuRequest) (*systemV1.Menu, error) {
	return s.uc.Create(ctx, req)
}

func (s *MenuService) UpdateMenu(ctx context.Context, req *systemV1.UpdateMenuRequest) (*systemV1.Menu, error) {
	ret, err := s.uc.Update(ctx, req)
	if err != nil {

		return nil, err
	}

	return ret, nil
}

func (s *MenuService) DeleteMenu(ctx context.Context, req *systemV1.DeleteMenuRequest) (*emptypb.Empty, error) {
	_, err := s.uc.Delete(ctx, req)
	if err != nil {

		return nil, err
	}

	return &emptypb.Empty{}, nil
}

File: frontend/apps/admin/src/rpc/service/menu.ts
import type { MenuService } from '#/rpc/api/admin/service/v1/i_menu.pb';
import type { Empty } from '#/rpc/api/google/protobuf/empty.pb';
import type { PagingRequest } from '#/rpc/api/pagination/v1/pagination.pb';
import type {
  CreateMenuRequest,
  DeleteMenuRequest,
  GetMenuRequest,
  ListMenuResponse,
  Menu,
  UpdateMenuRequest,
} from '#/rpc/api/system/service/v1/menu.pb';

import { MenuType } from '#/rpc/api/system/service/v1/menu.pb';
import { requestClient } from '#/rpc/request';

/** 后台菜单管理服务 */
class MenuServiceImpl implements MenuService {
  async CreateMenu(request: CreateMenuRequest): Promise<Empty> {
    return await requestClient.post<Empty>('/menus', request);
  }

  async DeleteMenu(request: DeleteMenuRequest): Promise<Empty> {
    return await requestClient.delete<Empty>(`/menus/${request.id}`);
  }

  async GetMenu(request: GetMenuRequest): Promise<Menu> {
    return await requestClient.get<Menu>(`/menus/${request.id}`);
  }

  async ListMenu(request: PagingRequest): Promise<ListMenuResponse> {
    return await requestClient.get<ListMenuResponse>('/menus', {
      params: request,
    });
  }

  async UpdateMenu(request: UpdateMenuRequest): Promise<Empty> {
    const id = request.menu?.id;
    if (request.menu !== null) request.menu.id = undefined;
    return await requestClient.put<Empty>(`/menus/${id}`, request);
  }
}

export const defMenuService = new MenuServiceImpl();

export const menuTypeList = [
  { value: MenuType.FOLDER, label: '目录' },
  { value: MenuType.MENU, label: '菜单' },
  { value: MenuType.BUTTON, label: '按钮' },
];
